### PR TITLE
Kernel+Userland: Implement 'wxallowed' mount option

### PR DESCRIPTION
**Pull request by jakewestrip**
This PR implements a 'wxallowed' mount option similar to OpenBSD's. The 'wxallowed' mount option allows any program located in a file system mounted with this option to violate w^x memory protection.

This will allow the user to run ports which violate the w^x protection (eg. for JIT code) without compromising any base system programs.

### CODE REVIEW

File: Kernel/Syscalls/mmap.cpp [hunk 1]
```diff
@@ -69,8 +69,12 @@ static bool should_make_executable_exception_for_dynamic_loader(bool make_readab
     return true;
 }
 
-static bool validate_mmap_prot(int prot, bool map_stack, bool map_anonymous, Memory::Region const* region = nullptr)
+static bool validate_mmap_prot(const Process& process, int prot, bool map_stack, bool map_anonymous, Memory::Region const* region = nullptr)
 {
+    bool wxallowed = process.executable()->mount_flags() & MS_WXALLOWED;
+    if (wxallowed)
+        return true;
```
Comment (kleinesfilmroellchen) [inline Kernel/Syscalls/mmap.cpp:76]: This means we allow other forbidden protection combinations as well, like executable&stack or executable&anonymous.  We most definitely don't want that :^). The check should probably go before line 81 (85) to skip the W^X check if we have wxallowed.

File: Kernel/Syscalls/mmap.cpp [hunk 2]
```diff
@@ -69,8 +69,12 @@ static bool should_make_executable_exception_for_dynamic_loader(bool make_readab
     return true;
 }
 
-static bool validate_mmap_prot(int prot, bool map_stack, bool map_anonymous, Memory::Region const* region = nullptr)
+static bool validate_mmap_prot(const Process& process, int prot, bool map_stack, bool map_anonymous, Memory::Region const* region = nullptr)
```
Comment (kleinesfilmroellchen) [inline Kernel/Syscalls/mmap.cpp:72]: This function can be moved into Process if we need access to it anyways.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Kernel/Syscalls/mmap.cpp (hunk 1 above)
Comment (jakewestrip) [reply Kernel/Syscalls/mmap.cpp]: Good catch, I've moved so that it just skips w^x checks

File: Kernel/Syscalls/mmap.cpp (hunk 2 above)
Comment (jakewestrip) [reply Kernel/Syscalls/mmap.cpp]: Moved both validation methods which use Process into Process (validate_mmap_prot and validate_inode_mmap_prot)

**Comment by kleinesfilmroellchen** [comment]
@jakewestrip please mark resolved comments as "resolved" so I know that you worked on it.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Please squash your fix commits with the original introduction of the code; it doesn't make sense to add code in one place just to move that code around in the next commit. This also goes for refactoring newly-written functions with new arguments etc.

File: Kernel/Syscalls/mmap.cpp [hunk 3]
```diff
@@ -333,11 +331,9 @@ ErrorOr<FlatPtr> Process::sys$mprotect(Userspace<void*> addr, size_t size, int p
...
                 return EPERM;
-            if (!validate_mmap_prot(prot, region->is_stack(), region->vmobject().is_anonymous(), region))
-                return EINVAL;
-            if (region->vmobject().is_inode()
-                && !validate_inode_mmap_prot(*this, prot, static_cast<Memory::InodeVMObject const&>(region->vmobject()).inode(), region->is_shared())) {
-                return EACCES;
+            TRY(validate_mmap_prot(prot, region->is_stack(), region->vmobject().is_anonymous(), region));
+            if (region->vmobject().is_inode()) {
+                TRY(validate_inode_mmap_prot(prot, static_cast<Memory::InodeVMObject const&>(region->vmobject()).inode(), region->is_shared()));
             }
```
Comment (kleinesfilmroellchen) [inline Kernel/Syscalls/mmap.cpp:337]: Remove curlies around this single-line if and all the others that you're touching

**Review by kleinesfilmroellchen — APPROVED** [review]

**Pull request merged**